### PR TITLE
feat: expose endpoint to delete ingested APIs

### PR DESCRIPTION
**Pull request by JedrzejJanasiak**
## Issue

https://gravitee.atlassian.net/browse/APIM-4894

## Description

PR with a new use case for deleting all the APIs previously ingested by this integration.

According the ACs the only APIs that are not deleted are these with PUBLISHED status 

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-vcwhjdwodm.chromatic.com)

### CODE REVIEW

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-apis.yaml [hunk 1]
```diff
@@ -1204,6 +1204,21 @@ paths:
...
                 default:
                     $ref: "#/components/responses/Error"
+        delete:
+            tags:
+                - Integrations
+            summary: Delete ingested APIs associated to integration
+            description: |-
+                Delete APIs that were ingested from 3rd party provider using this integration. 
+
+                To delete APIs user must have ENVIRONMENT_INTEGRATION[DELETE] permission.
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-apis.yaml]: The permission should be `ENVIRONMENT_API[DELETE]`

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/metadata/crud_service/MetadataCrudService.java [hunk 2]
```diff
@@ -23,4 +23,5 @@ public interface MetadataCrudService {
     Metadata create(Metadata metadata);
     Optional<Metadata> findById(MetadataId id);
     Metadata update(Metadata metadata);
+    void delete(Metadata metadata);
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/metadata/crud_service/MetadataCrudService.java]: Could you use `MetadataId` model instead the actual metadata model?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/integration/use_case/DeleteIngestedApisUseCase.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                .findActiveSubscriptionsByPlan(plan.getId())
+                .forEach(activeSubscription -> closeSubscriptionDomainService.closeSubscription(activeSubscription.getId(), auditInfo))
+        );
+
+        //Delete all subscriptions
+        plans.forEach(plan ->
+            subscriptionQueryService
+                .findSubscriptionsByPlan(plan.getId())
+                .forEach(subscription -> deleteSubscriptionDomainService.delete(subscription, auditInfo))
+        );
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/integration/use_case/DeleteIngestedApisUseCase.java]: What about merge the 3 foreach on plans?


```java
plans.forEach(plan -> {
            subscriptionQueryService
                .findActiveSubscriptionsByPlan(plan.getId())
                .forEach(activeSubscription -> closeSubscriptionDomainService.closeSubscription(activeSubscription.getId(), auditInfo));
        
        subscriptionQueryService
        .findSubscriptionsByPlan(plan.getId())
        .forEach(subscription -> deleteSubscriptionDomainService.delete(subscription, auditInfo));
        }

        deletePlanDomainService.delete(plan, auditInfo);
        );
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/membership/domain_service/DeleteMembershipDomainService.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        MembershipQueryService membershipQueryService,
+        MembershipCrudService membershipCrudService,
+        AuditDomainService auditDomainService
+    ) {
+        this.membershipQueryService = membershipQueryService;
+        this.membershipCrudService = membershipCrudService;
+        this.auditDomainService = auditDomainService;
+    }
+
+    public void delete(String apiId, AuditInfo auditInfo) {
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/membership/domain_service/DeleteMembershipDomainService.java]:
```suggestions
public void deleteApiMemberships(String apiId, AuditInfo auditInfo) {
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-apis.yaml (hunk 1 above)
Comment (JedrzejJanasiak) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/openapi/openapi-apis.yaml]: good catch

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/metadata/crud_service/MetadataCrudService.java (hunk 2 above)
Comment (JedrzejJanasiak) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/metadata/crud_service/MetadataCrudService.java]: sure, good idea

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/integration/use_case/DeleteIngestedApisUseCase.java (hunk 3 above)
Comment (JedrzejJanasiak) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/integration/use_case/DeleteIngestedApisUseCase.java]: 👍

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/membership/domain_service/DeleteMembershipDomainService.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        this.membershipQueryService = membershipQueryService;
+        this.membershipCrudService = membershipCrudService;
+        this.auditDomainService = auditDomainService;
+    }
+
+    public void deleteApiMemberships(String apiId, AuditInfo auditInfo) {
+        var memberships = membershipQueryService.findByReference(Membership.ReferenceType.API, apiId);
+        memberships.forEach(membership -> {
+            createAuditLog(auditInfo, membership, apiId);
+            membershipCrudService.delete(membership.getId());
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/membership/domain_service/DeleteMembershipDomainService.java]: I would flip the 2 instructions to create the audit when the deletion succeed

**Review by jgiovaresco — APPROVED** [review]

**Pull request merged**